### ISSUE TITLE
[HOLD FOR #23546][$500] mWeb - PDF - When typing password, revealing it and confirm, another attachment is displayed

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#35197]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.32-2
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4238270
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

## Action Performed:
Pre-requisite: the user must be logged in.
1. Go to any chat.
2. Send a password protected PDF.
3. Send another attachment.
4. Tap on the password protected PDF preview.
5. Type the password.
6. Tap on the eye icon button to reveal it (Important step).
7. Tap on "Confirm".

## Expected Result:
The password protected PDF should be displayed.

## Actual Result:
Another file is displayed.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [x] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/Expensify/App/assets/93399543/b1fd3194-5749-4f76-9086-6b01bfdc418e

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~014ec322edab15924f</li>
        <li>Upwork Job ID: 1750639602223542272</li>
        <li>Last Price Increase: 2024-02-08</li>
    </ul>
</details>

**Comment by kbecciv** [comment]
We think that this bug might be related to #vip-vsb
CC @quinthar

**Comment by jeremy-croff** [comment]
Seems like a side effect of https://github.com/Expensify/App/issues/23546
I noticed it diddn't even open the correct image on initial load.

**Comment by ikevin127** [comment]
# Proposal

### Please re-state the problem that we are trying to solve in this issue

Another file is displayed after inputting the password for a password protected pdf file, tapping the reveal password "Eye" then pressing the Confirm.

### What is the root cause of that problem?

I found the culprit, the line which for some weird unknown reason causes this issue, only after tapping the reveal password "Eye" then pressing the Confirm button. It was introduced by this PR https://github.com/Expensify/App/pull/26025 and this is the line:

https://github.com/Expensify/App/blob/1439ac08245bc7a49fc5b0384e53388155bdf407/src/components/PDFView/PDFPasswordForm.js#L145-L146

I know this is not enough to pass as an acceptable RCA since the reason why it's only happening when the reveal password "Eye" is tapped, is still unknown. But since I found a solution to the issue I thought I should mention it 🌞

### What changes do you think we should make in order to solve the problem?

Removing the line mentioned above fixes the issue but that wound also mean removing the functionality added by the mentioned PR which is: keyboard should stay open (keeping input focused) when clicking the Confirm button.

If we really want this fixed even though the full root cause is unknown, and we're fine with not having the "keyboard should stay open" functionality on iOS: mWeb Safari - we can create a library that will target only iOS: mWeb Safari, keeping the PR's functionality for all other platforms while fixing the current issue in Safari.

### Videos

<details>
<summary>iOS: mWeb Safari</summary>

| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/Expensify/App/assets/56457735/ec212331-423a-4b97-8ccd-95485c6fc160"> | <video src="https://github.com/Expensify/App/assets/56457735/f1852af8-ba77-43f6-ab87-cb9d71da5ea3"> |

</details>

**Comment by c3024** [comment]
Thanks for digging.

That did work. Let's see if we can identify the root cause.

**Comment by jeremy-croff** [comment]
I also had identified the mouseEvent tergerring the wrong image. And after this not making sense identified that initially the correct image did not even op in the video. The mouse event I believe is just a side effect of the original safari bug and the FlatList component

**Comment by ikevin127** [comment]
Indeed, @kidroca already identified a similar issue and raised it here: https://github.com/necolas/react-native-web/issues/2544, more details: https://github.com/Expensify/App/issues/23546#issuecomment-1895677762

From the react-native-web issue:

> When opening a Modal that includes a ScrollView or a VirtualList spanning multiple pages, the ScrollView automatically scrolls to the first element. This issue arises from the modal focus trap logic, even when the initialScrollIndex is set to a value other than the index of the first element.

For sure part of the root cause here is related to the FlatList edge case experienced on Safari - the modal's focus trap logic.

We just don't know exactly why does the reveal password together with the preventDefault on the onMouseDown of thr Confirm button produces the current issue.

**Comment by sourcecodedeveloper** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When typing password and clicking on the eye icon to reveal and then clicking confirm button, opens another attachment.

### What is the root cause of that problem?
The safari desktop and mobile browser are affected where clicking on the eye icon to reveal the password and then clicking confirm button, displays another pdf rather than the pdf for which the password was entered.

PR (https://github.com/Expensify/App/pull/26025) added e.preventDefault() to keep the keyboard open when an incorrect password was entered.
https://github.com/Expensify/App/blob/2e6e552aa9027d743752fdee3b0a8f92a6988940/src/components/PDFView/PDFPasswordForm.js#L146

Also the BaseTextInput has similar handling for onMouseDown
https://github.com/Expensify/App/blob/2e6e552aa9027d743752fdee3b0a8f92a6988940/src/components/TextInput/BaseTextInput/index.tsx#L402

If e.preventDefault were to be removed from BaseTextInput, then the keyboard would not stay open when the user clicks on the reveal password icon. The expected is to keep the keyboard open when the user user clicks on the reveal password icon.

So reverting the PR would also require consideration for the BaseTextInput onMouseDown event to be cleared but then the issue would get worse than before i.e the keyboard would be hidden if the user clicks on the eye icon or enters incorrect password.

### What changes do you think we should make in order to solve the problem?
I suggest not to revert the PR given that similar is used in BaseTextInput for onMouseDown event and reverting the PR would make the problem worse than before as the BaseTextInput behavior would also be affected.

Rather the simple approach would to patch the patch for Safari. The patch for safari involves checking the viewableItems array in the updatePage callback for onViewableItemsChanged. 

https://github.com/Expensify/App/blob/6198343741240266a259676c84552488151a2418/src/components/Attachments/AttachmentCarousel/index.js#L75

Due to the onMouseDown () => e.preventDefault, the viewabltItems array has the wrong item after the user clicks the eye to reveal the password and clicks confirm. 
The patch would involve to check the index of the viewable items array and if the index does not match with current page index, then call scrollToIndex passing current page index. Call to scrollToIndex loads the correct item in the viewable items array with the correct index.

```
            if(Browser.getBrowser() === CONST.BROWSER.SAFARI)
            {
                if(viewableItems.length === 1 && viewableItems[0].index !== pageRef.current)
                {
                    scrollRef.current.scrollToIndex({index: pageRef.current, animated: canUseTouchScreen});
                    return;
                }
            }
```


### What alternative solutions did you explore? (Optional)

**Comment by CortneyOfstad** [comment]
@c3024 bump on the proposals above? Thanks!

**Comment by c3024** [comment]
Thanks for the bump. I'll update in a few hours.

**Comment by c3024** [comment]
[Asked here on Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1707139186575409)

**Comment by sourcecodedeveloper** [comment]
@c3024 I wanted to mention the [proposal](https://github.com/Expensify/App/issues/35197#issuecomment-1915497108) also corrects the image not getting loaded correctly when more than one image is loaded, in desktop safari incorrect image is loaded. Also removing the onMouseDown event from PR(https://github.com/Expensify/App/pull/26025) would leave regression open as already mentioned and the BaseTextInput is also using onMouseDown. So to sum up, the issue affects mweb and desktop safari. the browser specific patch also leaves out any regression as already mentioned.

**Comment by c3024** [comment]
> @c3024 I wanted to mention the [proposal](https://github.com/Expensify/App/issues/35197#issuecomment-1915497108) also corrects the image not getting loaded correctly when more than one image is loaded, in desktop safari incorrect image is loaded. 
I think you mean this solution fixes another bug as well. Could you explain this preferably with a video? Thanks. @sourcecodedeveloper

**Comment by ikevin127** [comment]
> Also removing the onMouseDown event from PR(https://github.com/Expensify/App/pull/26025) would leave regression open as already mentioned and the BaseTextInput is also using onMouseDown. So to sum up, the issue affects mweb and desktop safari.

@c3024 The quote from above does not make any sense, the only tradeoff from my solution would be the keyboard persist which the mentioned PR added, that's it - no other regressions or BaseTextInput related issues. Here are the before / after videos on both mWeb and Web Safari.

### Videos

<details>
<summary>iOS: mWeb Safari</summary>

| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/Expensify/App/assets/56457735/ec212331-423a-4b97-8ccd-95485c6fc160"> | <video src="https://github.com/Expensify/App/assets/56457735/f1852af8-ba77-43f6-ab87-cb9d71da5ea3"> |

</details>

<details>
<summary>MacOS: Safari</summary>

| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/Expensify/App/assets/56457735/53e545c2-7f4a-4c6a-a8ac-043f3817b476"> | <video src="https://github.com/Expensify/App/assets/56457735/27046699-18f2-46e8-a2e1-038673bfeffc"> |

</details>

**Comment by sourcecodedeveloper** [comment]
@c3024 Yes the proposal resolves another bug. The safari bug happens without using passworded pdf, the incorrect image is loaded, when several pdfs are uploaded.
Safari bug

[safari_bug.webm](https://github.com/Expensify/App/assets/141340971/32cc9fd4-2a77-4f86-abe3-77d743279c62)

Safari Poc. The proposal resolves the issue of loading correct attachment.

[safari_poc.webm](https://github.com/Expensify/App/assets/141340971/3bdbd2f0-add1-442f-9b95-74c6f569f481)

**Comment by ikevin127** [comment]
The 'safari desktop issue of loading incorrect image' is already being handled here:
[[$2000] Wrong Image opens on HT Report in Safari](https://github.com/Expensify/App/issues/23546).

**Comment by CortneyOfstad** [comment]
@c3024 does this need to be closed in favor of https://github.com/Expensify/App/issues/23546?

**Comment by CortneyOfstad** [comment]
@c3024 bump on the question above ^^^ thanks!

**Comment by c3024** [comment]
I think we might want to fix this (though the fix does not completely address the root cause) independent of the other issue.

**Comment by sourcecodedeveloper** [comment]
@c3024 I am pretty sure the password pdf bug is due to the AttachmentCarousel loading incorrect page index which is why the incorrect pdf is loaded after entering the password. The onMouseDown approach does not completely address the root cause. The solution I have proposed would also resolve loading correct image as [mentioned](https://github.com/Expensify/App/issues/35197#issuecomment-1927252087). The best scenario going with the [proposed](https://github.com/Expensify/App/issues/35197#issuecomment-1915497108) safari solution would be the AttachmentCarousel would load the correct pdf after entering password and also the other safari bug would be resolved, and I do not have any worst scenario. If later the root cause for the onMouseDown event is discovered another PR can be raised, otherwise the issue would become blocked and come to stand still.

**Comment by c3024** [comment]
I think let’s just HOLD this till #23546 is fixed because it is the older and higher value issue and will likely address the root cause. Since #23546 appears to take a lot of time, it might be better to change this issue to `Weekly` or `Monthly`. @CortneyOfstad

**Comment by CortneyOfstad** [comment]
Sounds good @c3024 — I'll also update the title that it is on hold 👍

**Comment by CortneyOfstad** [comment]
No update on https://github.com/Expensify/App/issues/23546

**Comment by CortneyOfstad** [comment]
In the linked GH, there is discussion around next steps, so going to keep this at weekly until the PR is resolved.

**Comment by CortneyOfstad** [comment]
Bumped the other GH — will continue to keep an eye on things 👍

**Comment by CortneyOfstad** [comment]
Back from Ooo! Bumped the other GH and requested an update by EOD today. Will follow up here when received 👍

**Comment by CortneyOfstad** [comment]
Peter has found that the issue is affecting web browsers, so fix will be focused there 👍

**Comment by CortneyOfstad** [comment]
Bumped on the on-hold GH [here](https://github.com/Expensify/App/issues/23546#issuecomment-2034672966) — will update as soon as I hear back 👍

**Comment by CortneyOfstad** [comment]
Reached out for additional guidance/next steps [here](https://github.com/Expensify/App/issues/23546#issuecomment-2061263197)

**Comment by CortneyOfstad** [comment]
PR is in review now — https://github.com/Expensify/App/pull/39930

**Comment by CortneyOfstad** [comment]
PR is still in review 👍

**Comment by CortneyOfstad** [comment]
Bumped a comment on the PR [here](https://github.com/Expensify/App/pull/39930#issuecomment-2096163483)

**Comment by CortneyOfstad** [comment]
PR is continuing to be reviewed with comments as recent as <2 hours ago 👍

**Comment by CortneyOfstad** [comment]
PR has been reviewed and should be marked for merging based on the comment [here](https://github.com/Expensify/App/pull/39930#issuecomment-2111028261)

**Comment by CortneyOfstad** [comment]
We're on a merge freeze until Wednesday, so this will be delayed 👍

**Comment by CortneyOfstad** [comment]
Merge freeze should end today 🤞

**Comment by CortneyOfstad** [comment]
Went into product:ion 4 days ago! 😍

**Comment by CortneyOfstad** [comment]
On hold issue has been resolved/completed, so this can be closed! 🎉
